src/matrix/mod.rs

```diff
@@ -316,3 +316,80 @@ impl Stream for MatrixClient {
         self.poll_sync()
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::MatrixClient;
+    use mockito::{mock, Matcher, Matcher::UrlEncoded};
+    #[test]
+    fn matrix_login() {
+        let base_url = mockito::server_url().as_str().parse::<url::Url>().unwrap();
+        let username = "sample_username".to_string();
+        let password = "sample_password".to_string();
+        let mut core = tokio_core::reactor::Core::new().expect("could not create a tokio core");
+        let handle = core.handle();
+
+        let mock_req = mock("POST", "/_matrix/client/r0/login?")
+            .with_header("content-type", "application/json")
+            .with_status(200)
+            .create();
+
+        // run the future to completion. The future will error since invalid json is
+        // returned, but as long as the call is correct the error is outside the scope of this
+        // test. It is explicitly handled here in case the mock assert panics.
+        if let Err(e) = core.run(MatrixClient::login(
+            handle.clone(),
+            base_url,
+            username,
+            password,
+        )) {
+            println! {"MatrixSyncClient returned an error: {:?}", e}
+        }
+
+        // give the executor some time to execute the http request on a thread pool
+        std::thread::sleep(std::time::Duration::from_millis(200));
```

Comment:
I would be nice to get rid of this. I tried without the sleeps and it seems to work just fine, can you confirm that?

Reply:
This was an issue that happened when spawning the futures instead of using `core.run()`. I have removed it since we no longer spawn them.